python/tvm/octo/utils/target_info.py

```diff
@@ -0,0 +1,156 @@
+    if shutil.which("llc") is None:
+        return "llvm"
+
+    # Get host information from llc
+    stream = subprocess.run(["llc", "--version"], stdout=subprocess.PIPE)
+    cpu_info = stream.stdout.decode("utf-8")
+
+    # Parse out cpu line
+    cpu = re.search("(?<=Host CPU: ).+", cpu_info).group(0)
+
+    # Next extract attribute string.
+    platform = sys.platform
+    # Linux
+    if platform in ["linux", "linux2"]:
+        output = subprocess.check_output("lscpu", shell=True).decode()
+        pattern = r"^([^:]+):\s+(.*)$"
+        cpu_info = {}
+
+        for line in output.splitlines():
+            match = re.match(pattern, line)
+            if match:
+                key = match.group(1)
+                value = match.group(2)
+                cpu_info[key] = value.lower().strip()
+
+        features = cpu_info["Flags"].split(" ")
+        march = cpu_info["Architecture"]
+        cores = cpu_info["Core(s) per socket"]
+        sockets = cpu_info["Socket(s)"]
+        total_cores = str(int(cores) * int(sockets))
+        # Special case for x86_64 mismatch between underscore and hyphen
+        if march == "x86_64":
+            march = "x86-64"
+    else:
+        raise ValueError("Platform %s is not supported." % platform)
+
+    # Now we'll extract the architecture of the target.
+    output = subprocess.check_output("llc --version", shell=True).decode()
+    # Remove header.
+    march_options = re.search("(?<=Registered Targets:).*", output, re.DOTALL).group(0)
+    march_list = []
+    for march_line in march_options.split("\n"):
+        if march_line != "":
+            march_list.append(march_line.strip().split(" ")[0])
+    valid_march = False
+    if march in march_list:
+        valid_march = True
+
+    # Build the base target.
+    host_target = (
+        subprocess.check_output("llvm-config --host-target", shell=True).decode().strip("\n")
+    )
+    target = "llvm -mcpu=%s -mtriple=%s -num-cores=%s" % (cpu, host_target, total_cores)
+
+    # If possible, add more attribute information.
+    if valid_march:
+        # Get list of valid attributes for the target architecture.
+        attrs_info = subprocess.check_output(
+            "llc -march=%s -mattr=help" % march, shell=True, stderr=subprocess.STDOUT
+        ).decode()
+        supported_attrs = re.search(
+            r"(?<=Available features for this target:).*(?=Use \+feature to enable a feature)",
+            attrs_info,
+            re.DOTALL,
+        ).group(0)
+        attrs_list = []
+        for attrs_line in supported_attrs.split("\n"):
+            if attrs_line != "":
+                attrs_list.append(attrs_line.strip().split(" ")[0])
+
+        attrs = []
+        # Find which features are supported attrs.
+        for f in features:
+            if f in attrs_list:
+                attrs.append(f)
+
+        # Compuse attributes into valid string.
+        attrs_string = ""
```

Comment:
slightly cleaner:

attrs_string = ",".join(f"+{a}" for a in attrs)

Reply:
Also avoids the issue with the final comma below.

Fun fact: It's also more efficient (not that it's a concern here). `join` is generally faster than repeated concatenation in Python